File: nodejs/src/ui/chatbox.js
import { Branding } from './branding.js';

const C = Branding.colors;

const visible = (s) => s.replace(/\x1b\[[0-9;]*m/g, '').length;

function padRight(text, width) {
  const t = String(text ?? '');
  return t + ' '.repeat(Math.max(0, width - visible(t)));
}

function wrapLines(text, width) {
  const out = [];
  const src = String(text ?? '');
  for (const raw of src.split('\n')) {
    if (!raw.trim()) {
      out.push('');
      continue;
    }
    const words = raw.split(/\s+/).filter(Boolean);
    let cur = '';
    for (const word of words) {
      if (word.length >= width) {
        if (cur) {
          out.push(cur);
          cur = '';
        }
        for (let i = 0; i < word.length; i += width) out.push(word.slice(i, i + width));
        continue;
      }
      const next = cur ? `${cur} ${word}` : word;
      if (visible(next) > width) {
        if (cur) out.push(cur);
        cur = word;
      } else {
        cur = next;
      }
    }
    if (cur) out.push(cur);
  }
  return out;
}

function hardWrap(text, width) {
  const out = [];
  const src = String(text ?? '').replace(/\t/g, '  ');
  for (const raw of src.split('\n')) {
    if (visible(raw) <= width) {
      out.push(raw);
    } else {
      for (let i = 0; i < raw.length; i += width) out.push(raw.slice(i, i + width));
    }
  }
  return out;
}

// Column geometry: true-full width, then a centered column like opencode.
export function columnWidth(termWidth) {
  const w = termWidth && termWidth > 44 ? termWidth : 96;
  return Math.max(40, Math.min(100, w - 8));
}

export function centerPad(col, termWidth) {
  const w = termWidth && termWidth > 44 ? termWidth : 96;
  const pad = Math.max(0, Math.floor((w - col) / 2));
  return ' '.repeat(pad);
}

function highlight(text) {
  return String(text)
    .replace(/\*\*([^*]+)\*\*/g, `${C.bold}$1${C.reset}`)
    .replace(/`([^`]+)`/g, `${C.accent}$1${C.reset}`);
}

function splitBlocks(text) {
  const blocks = [];
  const re = /```([\w./+-]*)[^\n]*\n?([\s\S]*?)(?:```|$)/g;
  let last = 0;
  let m;
  while ((m = re.exec(String(text)))) {
    if (m.index > last) blocks.push({ code: false, text: text.slice(last, m.index) });
    blocks.push({ code: true, lang: m[1] || 'code', content: m[2] });
    last = m.index + m[0].length;
  }
  if (last < text.length) blocks.push({ code: false, text: text.slice(last) });
  return blocks;
}

export class FrameThrottle {
  constructor(render, min = 40) {
    this.render = render;
    this.min = min;
    this.last = 0;
  }

  schedule() {
    const now = performance.now();
    if (now - this.last >= this.min) {
      this.last = now;
      this.render();
    }
  }

  flush() {
    this.render();
  }
}

export class ChatUI {
  constructor({ model = '', engine = '' } = {}) {
    this.model = model;
    this.engine = engine;
    this.status = '';
    this.notices = [];
    this.streaming = null;
    this.thinking = false;
    this.compositor = null;
  }

  setCompositor(fn) {
    this.compositor = fn;
  }

  width() {
    const w = process.stdout.columns;
    return w && w > 44 ? w : 96;
  }

  setModel(model, engine) {
    this.model = model;
    this.engine = engine;
  }

  setStatus(text) {
    this.status = String(text ?? '');
  }

  setThinking(flag) {
    this.thinking = !!flag;
    this.streaming = null;
  }

  setStreaming(text) {
    this.streaming = String(text ?? '');
    this.thinking = false;
  }

  notify(text) {
    if (text) this.notices.push(String(text));
  }

  clearNotices() {
    this.notices = [];
  }

  _codeBox(lang, content, w) {
    const out = [];
    const c = C.success;
    // The box lives inside the centered column and carries its own 4-space
    // indent, so draw it at (w - 4) to stay inside the column.
    const box = Math.max(8, w - 4);
    const inner = Math.max(1, box - 8);
    const head = `    ${c}╭─${C.reset} ${C.bold}${c}${lang || 'code'}${C.reset}${c} `;
    const fill = Math.max(0, box - visible(head) - 1);
    out.push(`${head}${'─'.repeat(fill)}╮${C.reset}`);
    const lines = hardWrap(content, inner);
    while (lines.length && lines[lines.length - 1].trim() === '') lines.pop();
    if (!lines.length) lines.push('');
    for (const line of lines) {
      out.push(`    ${c}│${C.reset} ${padRight(line, inner)} ${c}│${C.reset}`);
    }
    out.push(`    ${c}╰${'─'.repeat(Math.max(0, box - 6))}╯${C.reset}`);
    return out;
  }

  _contentRows(content, w) {
    const rows = [];
    const width = Math.max(20, w - 4);
    const blocks = splitBlocks(content);
    for (const b of blocks) {
      if (b.code) {
        rows.push(...this._codeBox(b.lang, b.content, w));
      } else {
        const lines = wrapLines(b.text, width);
        if (lines.length === 0) lines.push('');
        for (const line of lines) {
          rows.push(line.trim() === '' ? '' : `    ${highlight(line)}`);
        }
      }
    }
    return rows;
  }

  _centered(row) {
    return `${centerPad(columnWidth(process.stdout.columns), process.stdout.columns)}${row}`;
  }

  // Banner drawn as part of the full clear+draw cycle so it can never be
  // wiped by a later render (the old showBanner() was erased instantly).
  _bannerRows(col) {
    const out = [];
    // Center the art as a BLOCK: pad every line to the widest one, then use a
    // single shared left indent so the llama's rows stay glued together.
    const art = Branding.BANNER.split('\n').map((l) => l.trimEnd());
    const artW = Math.max(...art.map((l) => visible(l)));
    const padL = Math.max(0, Math.floor((col - artW) / 2));
    for (const line of art) {
      out.push(`${' '.repeat(padL)}${C.primary}${line}${C.reset}`);
    }
    const center = (s) => {
      const p = Math.max(0, Math.floor((col - visible(s)) / 2));
      return `${' '.repeat(p)}${s}`;
    };
    out.push(center(`${C.bold}${C.accent}▸ ${Branding.APP_NAME} · ${this.model} · ${this.engine}${C.reset}`));
    out.push('');
    const box = Math.max(24, col - 6);
    const padR = (s) => s + ' '.repeat(Math.max(1, box - visible(s)));
    // Border is `  ┌` + (box+2) dashes + `┐` = box+6 wide, matching every row.
    const borderLine = `  ${C.primary}┌${'─'.repeat(box + 2)}┐${C.reset}`;
    out.push(center(borderLine));
    out.push(center(`  ${C.primary}│${C.reset} ${padR('  ' + C.bold + C.primary + 'Model' + C.reset)} ${C.primary}│${C.reset}`));
    out.push(center(`  ${C.primary}│${C.reset} ${padR('   ' + C.accent + this.model + C.reset)} ${C.primary}│${C.reset}`));
    out.push(center(`  ${C.primary}│${C.reset} ${padR('  ' + C.bold + C.primary + 'Engine' + C.reset)} ${C.primary}│${C.reset}`));
    out.push(center(`  ${C.primary}│${C.reset} ${padR('   ' + C.accent + this.engine + C.reset)} ${C.primary}│${C.reset}`));
    out.push(center(`  ${C.primary}└${'─'.repeat(box + 2)}┘${C.reset}`));
    out.push('');
    out.push(center(`${C.dim}Type /help for commands · /quit to exit${C.reset}`));
    out.push('');
    return out;
  }

  render(messages) {
    const w = this.width();
    const col = columnWidth(w);
    const out = [];
    out.push(`\x1b[2J\x1b[H`);
    // Banner (opencode style header) at the top of the centered column.
    out.push(...this._bannerRows(col));
    if (this.status) out.push(this._centered(`  ${C.dim}${this.status}${C.reset}`));
    out.push('');

    for (const m of messages || []) {
      if (!m || m.hidden || !m.content) continue;
      const title = m.role === 'user' ? 'You' : Branding.AI_PROMPT;
      const color = m.role === 'user' ? C.accent : C.primary;
      out.push(this._centered(`  ${C.bold}${color}${title}${C.reset}`));
      for (const row of this._contentRows(m.content, col)) {
        out.push(this._centered(row));
      }
      out.push('');
    }

    if (this.thinking) {
      out.push(this._centered(`  ${C.bold}${C.primary}${Branding.AI_PROMPT}${C.reset}`));
      out.push(this._centered(`    ${C.dim}···${C.reset}`));
      out.push('');
    } else if (this.streaming !== null) {
      out.push(this._centered(`  ${C.bold}${C.primary}${Branding.AI_PROMPT}${C.reset}`));
      const rows = this._contentRows(this.streaming, col);
      for (const r of rows.length ? rows : ['    ']) out.push(this._centered(r));
      out.push('');
    }

    for (const n of this.notices) {
      for (const line of wrapLines(n, col - 4)) out.push(this._centered(`  ${C.dim}${line}${C.reset}`));
    }
    if (this.notices.length) out.push('');

    // Keep the latest content pinned to the bottom of the terminal so streaming
    // never re-scrolls through older messages (which caused a flash).
    const isTty = process.stdout.isTTY && process.stdout.rows > 8;
    const rows = isTty ? process.stdout.rows : 999;
    const reserved = this.compositor ? Math.max(0, this.compositor.reservedRows()) : 0;
    const available = Math.max(3, rows - reserved);
    const lines = out.join('\n').split('\n').slice(1);
    const body = lines.length > available ? lines.slice(lines.length - available) : lines;
    process.stdout.write('\x1b[2J\x1b[H' + body.join('\n') + '\n');
    if (this.compositor) this.compositor.draw();
  }
}

File: python/vierrataleai/ui/chatbox.py
import re

from .branding import Branding

C = Branding.COLORS
_BOLD = "\033[1m"
_RESET = "\033[0m"
_DIM = "\033[2m"
_CYAN = "\033[38;2;6;182;212m"
_GREEN = "\033[38;2;16;185;129m"
_PURPLE = "\033[38;2;124;58;237m"
_CLEAR = "\x1b[2J\x1b[H"


def _visible(s: str) -> int:
    return len(re.sub(r"\x1b\[[0-9;]*m", "", s))


def _pad_right(text: str, width: int) -> str:
    t = str(text or "")
    return t + " " * max(0, width - _visible(t))


def _wrap_lines(text: str, width: int) -> list:
    out = []
    src = str(text or "")
    for raw in src.split("\n"):
        if not raw.strip():
            out.append("")
            continue
        words = [w for w in raw.split() if w]
        cur = ""
        for word in words:
            if len(word) >= width:
                if cur:
                    out.append(cur)
                    cur = ""
                for i in range(0, len(word), width):
                    out.append(word[i : i + width])
                continue
            nxt = f"{cur} {word}" if cur else word
            if _visible(nxt) > width:
                if cur:
                    out.append(cur)
                cur = word
            else:
                cur = nxt
        if cur:
            out.append(cur)
    return out


def _hard_wrap(text: str, width: int) -> list:
    out = []
    src = str(text or "").replace("\t", "  ")
    for raw in src.split("\n"):
        if _visible(raw) <= width:
            out.append(raw)
        else:
            for i in range(0, len(raw), width):
                out.append(raw[i : i + width])
    return out


def _column_width(term_width: int) -> int:
    w = term_width if term_width and term_width > 44 else 96
    return max(40, min(100, w - 8))


def _center_pad(col: int, term_width: int) -> str:
    w = term_width if term_width and term_width > 44 else 96
    pad = max(0, (w - col) // 2)
    return " " * pad


def _highlight(text: str) -> str:
    def bold(m):
        return f"{_BOLD}{m.group(1)}{_RESET}"

    def code(m):
        return f"{_CYAN}{m.group(1)}{_RESET}"

    text = re.sub(r"\*\*([^*]+)\*\*", bold, str(text))
    return re.sub(r"`([^`]+)`", code, text)


def _split_blocks(text: str) -> list:
    blocks = []
    pattern = re.compile(r"```([\w./+-]*)[^\n]*\n?([\s\S]*?)(?:```|$)")
    last = 0
    for m in pattern.finditer(str(text)):
        if m.start() > last:
            blocks.append({"code": False, "text": text[last : m.start()]})
        blocks.append(
            {"code": True, "lang": m.group(1) or "code", "content": m.group(2)}
        )
        last = m.end()
    if last < len(text):
        blocks.append({"code": False, "text": text[last:]})
    return blocks


def _code_box(lang: str, content: str, w: int) -> list:
    """Render a code box at width *w* (already inside the centered column)."""
    out = []
    # The box carries its own 4-space indent, so shrink the horizontal
    # dimensions so the whole box fits inside the column.
    box = max(8, w - 4)
    inner = max(1, box - 8)
    head = f"    {_GREEN}╭─{_RESET} {_BOLD}{_GREEN}{lang or 'code'}{_RESET}{_GREEN} "
    fill = max(0, box - _visible(head) - 1)
    out.append(f"{head}{'─' * fill}╮{_RESET}")
    lines = _hard_wrap(content, inner)
    while lines and lines[-1].strip() == "":
        lines.pop()
    if not lines:
        lines.append("")
    for line in lines:
        out.append(f"    {_GREEN}│{_RESET} {_pad_right(line, inner)} {_GREEN}│{_RESET}")
    out.append(f"    {_GREEN}╰{'─' * max(0, box - 6)}╯{_RESET}")
    return out


def _content_rows(content: str, w: int) -> list:
    rows = []
    width = max(20, w - 4)
    for b in _split_blocks(content):
        if b["code"]:
            rows.extend(_code_box(b["lang"], b["content"], w))
        else:
            lines = _wrap_lines(b["text"], width)
            if not lines:
                lines = [""]
            for line in lines:
                rows.append("" if line.strip() == "" else f"    {_highlight(line)}")
    return rows


class FrameThrottle:
    def __init__(self, render):
        self._render = render
        self._last = 0.0

    def schedule(self):
        import time

        now = time.monotonic()
        if now - self._last >= 0.04:
            self._last = now
            self._render()

    def flush(self):
        self._render()


def _centered(row: str, pad: str) -> str:
    return f"{pad}{row}"


class ChatUI:
    def __init__(self, model: str = "", engine: str = ""):
        self.model = model
        self.engine = engine
        self.status = ""
        self.notices = []
        self.streaming = None
        self.thinking = False
        self.compositor = None

    def set_compositor(self, compositor) -> None:
        self.compositor = compositor

    def width(self):
        import shutil

        try:
            return shutil.get_terminal_size().columns
        except Exception:
            return 96

    def set_model(self, model: str, engine: str) -> None:
        self.model = model
        self.engine = engine

    def set_status(self, text: str) -> None:
        self.status = str(text or "")
        self.thinking = False
        self.streaming = None

    def set_thinking(self, flag: bool) -> None:
        self.thinking = bool(flag)
        self.streaming = None

    def set_streaming(self, text: str) -> None:
        self.streaming = str(text or "")
        self.thinking = False

    def notify(self, text: str) -> None:
        if text:
            self.notices.append(str(text))

    def clear_notices(self) -> None:
        self.notices = []

    def _banner_rows(self, col):
        """Banner drawn as part of the full clear+draw cycle so the old
        show_banner() (erased by the first render) can't flash/garbled."""
        out = []

        # Center the art as a BLOCK: pad every line to the widest one, then use
        # a single shared left indent so the llama's rows stay glued together.
        art = [line.rstrip() for line in Branding.BANNER.split("\n")]
        art_w = max(_visible(line) for line in art) if art else 0
        pad_l = max(0, (col - art_w) // 2)
        for line in art:
            out.append(f"{' ' * pad_l}{_PURPLE}{line}{_RESET}")

        def center(s):
            return " " * max(0, (col - _visible(s)) // 2) + s
        out.append(
            center(
                f"{_BOLD}{_CYAN}▸ {Branding.APP_NAME} · {self.model} · {self.engine}{_RESET}"
            )
        )
        out.append("")
        box = max(24, col - 6)

        def pad_r(s):
            return s + " " * max(1, box - _visible(s))

        row = lambda content: center(f"  {_PURPLE}│{_RESET} {content} {_PURPLE}│{_RESET}")
        out.append(center(f"  {_PURPLE}┌{'─' * (box + 2)}┐{_RESET}"))
        out.append(row(pad_r(f"  {_BOLD}{_PURPLE}Model{_RESET}")))
        out.append(row(pad_r(f"   {_CYAN}{self.model}{_RESET}")))
        out.append(row(pad_r(f"  {_BOLD}{_PURPLE}Engine{_RESET}")))
        out.append(row(pad_r(f"   {_CYAN}{self.engine}{_RESET}")))
        out.append(center(f"  {_PURPLE}└{'─' * (box + 2)}┘{_RESET}"))
        out.append("")
        out.append(center(f"{_DIM}Type /help for commands · /quit to exit{_RESET}"))
        out.append("")
        return out

    def render(self, messages) -> None:
        import shutil
        import sys

        w = self.width()
        col = _column_width(w)
        pad = _center_pad(col, w)
        out = [_CLEAR]
        # Banner (opencode style header) at the top of the centered column.
        for row in self._banner_rows(col):
            out.append(f"{pad}{row}")
        if self.status:
            out.append(f"{pad}  {_DIM}{self.status}{_RESET}")
        out.append("")

        for m in messages or []:
            if not m or m.get("hidden") or not m.get("content"):
                continue
            title = "You" if m.get("role") == "user" else Branding.AI_PROMPT
            color = _CYAN if m.get("role") == "user" else _PURPLE
            out.append(_centered(f"  {_BOLD}{color}{title}{_RESET}", pad))
            for row in _content_rows(m.get("content"), col):
                out.append(_centered(row, pad))
            out.append("")

        if self.thinking:
            out.append(_centered(f"  {_BOLD}{_PURPLE}{Branding.AI_PROMPT}{_RESET}", pad))
            out.append(_centered(f"    {_DIM}···{_RESET}", pad))
            out.append("")
        elif self.streaming is not None:
            out.append(_centered(f"  {_BOLD}{_PURPLE}{Branding.AI_PROMPT}{_RESET}", pad))
            rows = _content_rows(self.streaming, col)
            for r in rows or ["    "]:
                out.append(_centered(r, pad))
            out.append("")

        for n in self.notices:
            for line in _wrap_lines(n, col - 4):
                out.append(_centered(f"  {_DIM}{line}{_RESET}", pad))
        if self.notices:
            out.append("")

        # Pin the latest content to the bottom so streaming never re-scrolls
        # through older messages (which caused a flash).
        rows = 999
        try:
            if sys.stdout.isatty():
                size = shutil.get_terminal_size().lines
                if size and size > 8:
                    rows = size
        except Exception:
            pass
        content = out[1:]
        reserved = 0
        if self.compositor is not None:
            reserved = max(0, self.compositor.reserved_rows())
        available = max(3, rows - reserved)
        body = content[-available:] if len(content) > available else content
        print(_CLEAR + "\n".join(body), flush=True)
        if self.compositor is not None:
            self.compositor.draw()